### DISCUSSION TITLE
[Maven] Handling datetime snapshot released dependencies

### DISCUSSION THREAD

**Discussion by setchy** [renovatebot/renovate#9982] [Request Help]
I'd like to ask this community for ideas on the best way to handle ignoring datetime snapshot dependencies that are released to Maven Central

An example is https://search.maven.org/artifact/com.graphql-java/graphql-java

For now I'm using the following `allowedVersions` regex and although it is functional, I'm curious if there is a better way to achieve this, or if an enhancement to the maven versioning logic would be beneficial
```
{
    "groupName": "GraphQL Java",
    "matchPackagePrefixes": ["com.graphql-java:"],
    "allowedVersions": "/^[0-9]+\\.[0-9]+(\\.[0-9]+)?$/"
},
```

Locally I added some additional tests to [versioning/maven/index.spec.ts](https://github.com/renovatebot/renovate/blob/main/lib/versioning/maven/index.spec.ts) and interestingly when passing `2021-05-03T08-53-12-8f1ec444` into the maven isStable function it returns true

**Comment by rarkins** [comment]
Is there any part of the Maven spec which says that type of version is unstable? If so then we could fix it within versioning

**Reply by klaasdellschaft** [reply]
Hi,

I have a very similar problem which is potentially related, and where Renovate (and potentially also the Maven specification how to compare version numbers) is not according to my intuition / to what is typically desired. 

### Problem

In my example, the Maven repository contains the following two version numbers:

- `3.8.1`
- `3.8.1-jboss-2`

If I use [Maven's tool for Version Order Testing](https://maven.apache.org/pom.html#version-order-testing), the two version numbers will compare as 

- `3.8.1` < `3.8.1-jboss-2`
 
I would have expected, that those version numbers are "incomparable" but unfortunately Maven is able to bring all version numbers into a strict order, even if it deals with qualifiers like `jboss` which are in my opinion not suited for being brought into an order. (Of course, you could argue that `jboss` should not be put into the version number but into the classifier field of Maven, and I would agree with you.)

### Where my problem touches the original topic of this post (and the follow-up question whether a version should be treated as unstable):

The [Maven versioning spec](https://maven.apache.org/pom.html#version-order-specification) contains among others the following sentence:

> Non-numeric tokens ("qualifiers") have the alphabetical order, except for the following tokens which come first in this order:
> "alpha" < "beta" < "milestone" < "rc" = "cr" < "snapshot" < "" = "final" = "ga" < "sp"

According to this definition, I would see everything that is having the qualifier "" (as in `3.8.1`, which is normalized to `3.8.1-""` when comparing to `3.8.1-jboss`) as stable. The same applies for the qualifiers "final" and "ga", as well as "sp" (which seems to abbreviate "service pack"). In contrast, everything "smaller" than that (i.e. "alpha", "beta", ...) I would see as unstable.

Unfortunately, all non-numeric qualifiers that are not in that special list are treated as "bigger" than a release, or even a service pack. Thus, in extension, one **may** infer that they are stable (in the end, they are "bigger" than a release) although it is not explicitly mentioned in the Maven spec.

### Applying this interpretation to the original posting

A version number like `2021-05-03T08-53-12-8f1ec444` is a mix between numerical and non-numerical tokens (e.g. the `T` would be treated as a non-numerical token, as well as the characters in the hash). It would be normalized as follows (based on the tool provided by Maven):

- 2021-05-03T08-53-12-8f1ec444 == 2021-5-3-t-8-53-12-8-f-1-ec-444

Based on my interpretation from above, I would see that as a stable version, which is definitely "bigger" than, e.g., `1.0.0` because 2021 > 1. However, with a bit of "bad luck" regarding hash value at the end, this may as well be treated as an "alpha" version. For example:

- 2021-05-03T08-53-12-8f1a1444 == 2021-5-3-t-8-53-12-8-f-1-**alpha**-1444

I.e., if the hash contains an "a" or "b" surrounded by numbers, this would make it to an `alpha` or `beta` version, which would be considered unstable.

### Summary

In summary, I would say that based on Maven's strict ordering of version numbers, unfortunately Renovate is most likely right in putting those versions into an unexpected / undesired order, and it is also most likely right in considering them as stable versions. 

However, given the undesired effects in Maven's version ordering, maybe some Renovate preset could be defined which would only allow numerical versions for Maven dependencies or (alternatively) the [Semantic Versioning 1.0.0](https://semver.org/spec/v1.0.0.html) subset of valid Maven version numbers?

Given that Renovate somewhere most likely already contains an implementation / check whether a version number is following Semantic Versioning 1.0.0, maybe instead of a preset with a potentially quite complicated regular expression for `allowedVersions`, it may also be an option to introduce a corresponding configuration option.
